### PR TITLE
Removes deprecated API usages and cleans up code

### PR DESCRIPTION
**Pull request by samagra14**
## Description

There has been a lot of deprecated API usages in the app. I have scanned the entire application for such deprecated APIs and have made corrections regarding the same. Besides, I have removed boilerplate code and made it efficient. The following are the list of APIs(deprecated) changed : -

| Existing API        | Replaced by           |
| ------------- |:-------------:| 
| `MenuItemCompat.setOnActionExpandListener()`    | `setOnActionExpandListener()` | 
| Java 7 Unambigous explicit types     | Generics      |   
| Not annotated parameters | Proper annotations     | 
|`lang3.StringEscapeUtils` | `text.StringEscapeUtils` |
|`MenuItemCompat.setShowAsAction()` | `static setShowAsAction()`|
|Anonymous Classes (jre <1.7)|lambdas|

I will be adding more commits to this PR but it is best to review simultaneously as the changes may accumulate. These changes are necessary because : 
* These hinder migration to MVVM architecture components
* Deprecated APIs have accumulated and if not upgraded now, then these will cause crashes in newer Android versions.
* These will make it easier to incorporate Kotlin in our workflow.
* Removes unnecessary code (366 - 279 = 87 lines less)

**I have been working on this for almost a week now and after scanning the code I have found a serious issue in the way we manage our Async tasks. Asynctasks in our app are non-static i.e they are nested objects within the activities. Now, if an activity gets destroyed while the execution of a background task is still pending, then the object of the Activity is not destroyed and there are references to the destroyed activity leading to memory leaks in almost every activity! This is a really serious issue and we should look into handling it.**

_Since this is a long PR I will request all the active members to review sections of the PR._

### CODE REVIEW

**Comment by samagra14** [comment]
@teolemon @Karljoones Please have a look.
@sachincool @Shubham2320 @ManavPoddar @jaztriumph @PrajwalM2212 @ManognaMutyala Can you please have a look and review the PR.

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/customtabs/CustomTabActivityHelper.java [hunk 1]
```diff
@@ -114,9 +114,8 @@ public boolean mayLaunchUrl(Uri uri, Bundle extras, List<Bundle> otherLikelyBund
         if (mClient == null) return false;
 
         CustomTabsSession session = getSession();
-        if (session == null) return false;
+        return session != null && session.mayLaunchUrl(uri, extras, otherLikelyBundles);
```
Comment (sachincool) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/views/customtabs/CustomTabActivityHelper.java]: cool one !!!

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AlertUserFragment.java [hunk 2]
```diff
@@ -162,18 +174,6 @@ protected void onAddAllergens() {
 
     }
 
-
-    public static Integer getKey(HashMap<Integer, String> map, String value) {
```
Comment (sachincool) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AlertUserFragment.java]: I couldn't see the alternate code for this method.

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java [hunk 3]
```diff
@@ -542,7 +537,7 @@ public void uploadOfflineImages(Context context, boolean cancel, JobParameters j
         if (!cancel) {
 //            Toast.makeText(context, "called function", Toast.LENGTH_SHORT).show();
             task.job = job;
-            task.service = new WeakReference<SavedProductUploadJob>(service);
+            task.service = new WeakReference<>(service);
```
Comment (sachincool) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java]: Was this necessary? doesn't this has any consequences?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/SwipeController.java [hunk 4]
```diff
@@ -91,62 +91,48 @@ public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHo
...
-                    else if (dX > buttonWidth) buttonShowedState = ButtonsState.GONE;
-
-                    if (buttonShowedState != ButtonsState.GONE) {
-                        setTouchDownListener(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
-                        setItemsClickable(recyclerView, false);
-                    }
+        recyclerView.setOnTouchListener((v, event) -> {
+            swipeBack = event.getAction() == MotionEvent.ACTION_CANCEL || event.getAction() == MotionEvent.ACTION_UP;
+            if (swipeBack) {
+                Log.e("setTouchListener", dX + "");
```
Comment (sachincool) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/SwipeController.java]: you could remove Log.e from there

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/SwipeControllerActions.java [hunk 5]
```diff
@@ -4,6 +4,7 @@
  * Created by Mehrosh.Mehboob on 20-Feb-18.
  */
 
+@FunctionalInterface
```
Comment (sachincool) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/SwipeControllerActions.java]: ik this one i read  java 9 :P

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AlertUserFragment.java (hunk 2 above)
Comment (ManognaMutyala) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AlertUserFragment.java]: He replace it o  line 51

Comment (sachincool) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AlertUserFragment.java]: my bad thanks @ManognaMutyala

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java (hunk 3 above)
Comment (samagra14) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java]: This was unnecessary, that's why I removed it. And it runs perfectly fine.

**Comment by teolemon** [comment]
Can you fix the conflicts @samagra14 ?

**Comment by samagra14** [comment]
@teolemon I have resolved the merge conflicts. Please merge this as soon as possible because there are a lot of changes across files and if not merged, conflicts may arise soon.

**Comment by samagra14** [comment]
@jinqian @Karljoones Please have a look here.

**Comment by Karljoones** [comment]
@samagra14 there are a lot of conflicts present in this PR, could you have a look into these before we review?

**Comment by Karljoones** [comment]
Poke @samagra14

**Comment by samagra14** [comment]
@Karljoones I am so sorry. Actually, I am busy with my exams (mid terms). They end tomorrow. I will finish the work by tomorrow evening.

**Comment by Karljoones** [comment]
Take your time @samagra14, concentrate on your exams for the time being!

**Comment by samagra14** [comment]
@Karljoones Please merge it as early as possible or else merge conflicts will arise again. I have already resolved conflicts twice. Please have a look.

**Comment by samagra14** [comment]
Ping @Karljoones.

**Pull request merged**